src/apps/shared/ordering.py

```diff
@@ -26,40 +63,136 @@ class Clause:
+        for value in args:
+            parsed_field = self._parse_ordered_field(value)
+            if parsed_field is None:
+                continue
+            direction, field = parsed_field
+            if not self._is_manual_field(field):
+                continue
+            parsed_fields.append((direction, self.manual_fields[field]))
+            # Flag if any of the fields cannot be sorted by SQL (i.e. manual only)
+            if not self._is_sql_field(field):
+                has_manual_only_fields = True
+        return parsed_fields if has_manual_only_fields else None
 
+    # Returns SQL clauses suitable for Query.order_by based on provided "+field"-style args
     def get_clauses(self, *args):
-        sorting_fields = []
+        clauses: list[str] = []
         for value in args:
-            ordering_field = self._prepare_ordered_field(value)
-            if ordering_field is not None:
-                sorting_fields.append(ordering_field)
-        return sorting_fields
+            parsed_field = self._parse_ordered_field(value)
+            if parsed_field is None:
+                continue
+            _direction, field = parsed_field
+            if not self._is_sql_field(field):
+                continue
+            clause = self._prepare_sql_clause(parsed_field)
+            if clause is not None:
+                clauses.append(clause)
+        return clauses
+
+    # Returns list of fields that this Ordering class supports for sorting, based on whether
+    # to use manual sorting or not. Returns field names in camelCase suitable for API output.
+    def get_ordering_fields(self, use_manual_sorting: bool = False):
+        use_manual_sorting = use_manual_sorting and bool(self.manual_fields)
+        fields = (self.manual_fields if use_manual_sorting else self.sql_fields).keys()
+        return list(to_camelcase(word) for word in fields)
+
+    def _is_sql_field(self, field: str):
+        return field in self.sql_fields
 
-    def _prepare_ordered_field(self, value: str):
+    def _is_manual_field(self, field: str):
+        return field in self.manual_fields
+
+    def _parse_ordered_field(self, value: str):
         if not value:
             return None
-        sign, field = value[0], value[1:]
-        if sign != "-":
+        direction, field = value[0], value[1:]
+        if direction != "-":
             field = value
-            sign = "+"
+            direction = "+"
         field = field.lstrip("+")
+        return (direction, field)
 
-        if field not in self.fields:
+    def _prepare_sql_clause(self, value: tuple[OrderingDirection, str]):
+        direction, field = value
+        if not self._is_sql_field(field):
             return None
 
-        return self.actions[sign](self.fields[field])
+        return self.actions[direction](self.sql_fields[field])
+
+    def manual_sort(self, data: list, fields: list[tuple[OrderingDirection, OrderingField]]):
```

Comment:
Don't you think that implementing this task would be easier using models and magic methods lt, gt, eq? The current solution will be hard to maintain.

Reply:
Great suggestion, and good point. Let me see how I can refactor this to keep type-specific comparison logic out of the Ordering class.